src/shims/windows/sync.rs

```diff
@@ -26,17 +30,112 @@ fn srwlock_get_or_create_id<'mir, 'tcx: 'mir>(
             // Caller of the closure needs to allocate next_id
             None
         } else {
-            Some(RwLockId::from_u32(old.to_u32().expect("layout is u32")))
+            Some(old.to_u32().expect("layout is u32"))
+        })
+    }
+
+    // Locks are pointer-sized pieces of data, initialized to 0.
+    // We use the first 4 bytes to store the RwLockId.
+    fn srwlock_get_or_create_id(
+        &mut self,
+        lock_op: &OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, RwLockId> {
+        let this = self.eval_context_mut();
+        let value_place = this.deref_operand_and_offset(lock_op, 0, this.machine.layouts.u32)?;
+
+        this.rwlock_get_or_create(|ecx, next_id| {
+            Ok(ecx.get_or_create_id(next_id.to_u32_scalar(), &value_place)?.map(RwLockId::from_u32))
+        })
+    }
+
+    // see above: condvars are implemented identically to locks
+    fn condvar_get_or_create_id(
+        &mut self,
+        condvar_op: &OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, CondvarId> {
+        let this = self.eval_context_mut();
+        let value_place = this.deref_operand_and_offset(condvar_op, 0, this.machine.layouts.u32)?;
+
+        this.condvar_get_or_create(|ecx, next_id| {
+            Ok(ecx
+                .get_or_create_id(next_id.to_u32_scalar(), &value_place)?
+                .map(CondvarId::from_u32))
         })
-    })
+    }
+
+    /// Try to reacquire the lock associated with the condition variable after we
+    /// were signaled.
+    fn reacquire_cond_lock(
+        &mut self,
+        thread: ThreadId,
+        lock: RwLockId,
+        shared: bool,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        this.unblock_thread(thread);
+
+        if shared {
```

Comment:
Please use "read" and "write" locked terminology, rather than "shared". That's what the rest of the codebase does AFAIK. I also find it much less confusing. :)